lib/AST/ModuleNameLookup.cpp

```diff
@@ -90,20 +95,111 @@ static bool updateOverloadSet(NamedCanTypeSet &overloads,
   return true;
 }
 
-/// After finding decls by name lookup, filter based on the given
-/// resolution kind and existing overload set and add them to \p results.
+/// Returns whether it's okay to stop recursively searching imports, given that
+/// we found something, are using a CanTypeSet to resolve overloads, and have
+/// the given \p resolutionKind.
+static bool canReturnEarly(const CanTypeSet &overloads,
+                           ResolutionKind resolutionKind) {
+  switch (resolutionKind) {
+  case ResolutionKind::Overloadable:
+    return false;
+  case ResolutionKind::Exact:
+  case ResolutionKind::TypesOnly:
+    return true;
+  }
+}
+
+/// Returns whether it's okay to stop recursively searching imports, given that
+/// we found something, are using a NamedCanTypeSet to resolve overloads, and
+/// have the given \p resolutionKind.
```

Comment:
My own style would be comment it by saying something like: "This function statically overrides the previous one to ensure that the result is always false for this argument type." But even that seems not "why" enough...